chacha20poly1305/src/impl_hacl.rs

```diff
@@ -0,0 +1,109 @@
+use crate::{AeadError, KEY_LEN, NONCE_LEN, TAG_LEN};
+
+/// The ChaCha20-Poly1305 AEAD encryption function. Writes the concatenation of the ciphertexoft
+/// produced by ChaCha20 and the MAC tag into `ctxt` and returns the two pieces separately.
+///
+/// This implementation is backed by hacl-rs and can only handle inputs up to a length of `u32::MAX`.
+/// When provided longer values, this function will return an error.
+pub fn encrypt<'a>(
+    key: &[u8; KEY_LEN],
+    ptxt: &[u8],
+    ctxt: &'a mut [u8],
+    aad: &[u8],
+    nonce: &[u8; NONCE_LEN],
+) -> Result<(&'a [u8], &'a [u8; TAG_LEN]), AeadError> {
+    let ptxt_len: u32 = ptxt
+        .len()
+        .try_into()
+        .map_err(|_| AeadError::PlaintextTooLarge)?;
+
+    let aad_len: u32 = aad.len().try_into().map_err(|_| AeadError::AadTooLarge)?;
+
+    if ctxt.len() < ptxt.len() + TAG_LEN {
+        return Err(AeadError::CiphertextTooShort);
+    }
+
+    // ensure destination slice has just the right length
+    let (ctxt_cpa, tag) = ctxt.split_at_mut(ptxt.len());
+    let tag: &mut [u8; TAG_LEN] = tag.try_into().unwrap();
+
+    crate::hacl::aead_chacha20poly1305::encrypt(
+        ctxt_cpa, tag, ptxt, ptxt_len, aad, aad_len, key, nonce,
+    );
+
+    Ok((ctxt_cpa, tag))
+}
+
+/// The ChaCha20-Poly1305 AEAD decryption function. Writes the result of the decryption to `ptxt`,
+/// and returns the slice of appropriate length.
```

Comment:
Why do you return the ptxt slice in addition to having it in `ptxt`?

Reply:
Because the slice I return is trimmed to the right size. The input slice might be longer.